## Finance 6320: Computational Finance <br>
**Daniel Wouden**
**Assignment #3** <br>
**Due: TBD** <br>



## Imports

In [6]:
import numpy as np
from scipy.stats import binom

**Problem 1.** 

In this problem you will use the classes that we have developed to price European puts and calls with the multiperiod Binomial Option Pricing Model (BOPM). This problem is more or less done for you in our lectures.

Demonstrate that this code works by replicating the pricing of the European put option in Figure 10.6 of the McDonald book on page 310. You should get a price of $\$2.999$ for the put option.

Also, price the corresponding call option to demonstrate the algorithm, implemented once, is polymorphic and is capable of pricing call options as well.

In [21]:
import numpy as np
from scipy.stats import binom

def callPayoff(spot, strike):
    return np.maximum(spot - strike, 0.0)

def putPayoff(spot, strike):
    return np.maximum(strike - spot, 0.0)

def EuropeanBinomialPricer(spot, strike, rate, vol, div, steps, expiry, payoff):
    h = expiry / steps
    nodes = steps + 1
    u = np.exp((rate - div) * h + vol * np.sqrt(h))
    d = np.exp((rate - div) * h - vol * np.sqrt(h))
    pstar = (np.exp((rate - div) * h) - d) / (u - d)
    disc = np.exp(-(rate - div))
    callT = 0.0
    
    for i in range(nodes):
        spotT = spot * (u ** (steps - i)) * (d ** i)
        callT += payoff(spotT, strike) * binom.pmf(steps - i, steps, pstar)
        
    callPrc = callT * disc
    return callPrc
spot = 41.0
strike = 40.0
vol = 0.3
rate = 0.08
div = 0.0
expiry = 1.0
steps = 3

answer = EuropeanBinomialPricer(spot, strike, rate, vol, div, steps, expiry, putPayoff)
## The Put option price is:
answer


2.9985071167431512

In [23]:
import numpy as np
from scipy.stats import binom

def callPayoff(spot, strike):
    return np.maximum(spot - strike, 0.0)

def putPayoff(spot, strike):
    return np.maximum(strike - spot, 0.0)

def EuropeanBinomialPricer(spot, strike, rate, vol, div, steps, expiry, payoff):
    h = expiry / steps
    nodes = steps + 1
    u = np.exp((rate - div) * h + vol * np.sqrt(h))
    d = np.exp((rate - div) * h - vol * np.sqrt(h))
    pstar = (np.exp((rate - div) * h) - d) / (u - d)
    disc = np.exp(-(rate - div))
    callT = 0.0
    
    for i in range(nodes):
        spotT = spot * (u ** (steps - i)) * (d ** i)
        callT += payoff(spotT, strike) * binom.pmf(steps - i, steps, pstar)
        
    callPrc = callT * disc
    return callPrc
spot = 41.0
strike = 40.0
vol = 0.3
rate = 0.08
div = 0.0
expiry = 1.0
steps = 3

answer = EuropeanBinomialPricer(spot, strike, rate, vol, div, steps, expiry, callPayoff)
## The Call Option price is:
answer


7.0738532612777147

**Problem 2.**

In this problem, building off of the code above, you will implement the American Binomial Option Pricing Model (ABOPM). Reprice both of the options above as American options. Outline the differences between the European and American solutions.

In [2]:
import numpy as np

class VanillaOption(object):
    def __init__(self, strike, expiry, payoff):
        self.__strike = strike
        self.__expiry = expiry
        self.__payoff = payoff
        
    @property
    def strike(self):
        return self.__strike
    
    @property
    def expiry(self):
        return self.__expiry
    
    def payoff(self, spot):
        return self.__payoff(self, spot)
        
def call_payoff(option, spot):
    return np.maximum(spot - option.strike, 0.0)

def put_payoff(option, spot):
    return np.maximum(option.strike - spot, 0.0)
    
def AmericanBinomial(option, spot, rate, volatility, dividend, steps):
    return np.maximum()
    
spot = 41
strike = 40.0
expiry = 1.0
rate = 0.08
volatility = 0.3
dividend = 0.0
steps = 3
num_nodes = steps + 1

dt = expiry / steps
u = np.exp((rate * dt) + volatility * np.sqrt(dt))
d = np.exp((rate * dt) - volatility * np.sqrt(dt))
pu = (np.exp(rate * dt) - d) / (u - d)
pd = 1 - pu
disc = np.exp(-rate * dt)
dpu = disc * pu
dpd = disc * pd
    
cashFlow_t = np.zeros((num_nodes, ))
spot_t = np.zeros((num_nodes, ))
option = VanillaOption(strike, expiry, put_payoff)
    
for i in range(num_nodes):
    spot_t[i] = spot * (u ** (steps - i)) * (d ** i)
    cashFlow_t[i] = option.payoff(spot_t[i]) 

for i in range((steps - 1), -1, -1):
    for j in range(i+1):
        cashFlow_t[j] = dpu * cashFlow_t[j] + dpd * cashFlow_t[j+1]
        spot_t[j] = spot_t[j] / u
        cashFlow_t[j] = np.maximum(cashFlow_t[j], option.payoff(spot_t[j]))    
    
price = cashFlow_t[0]

#The Put option price is:    
print ("The Option Price is: {0:.4f}".format(price))

The Option Price is: 3.2929


In [1]:
import numpy as np

class VanillaOption(object):
    def __init__(self, strike, expiry, payoff):
        self.__strike = strike
        self.__expiry = expiry
        self.__payoff = payoff
        
    @property
    def strike(self):
        return self.__strike
    
    @property
    def expiry(self):
        return self.__expiry
    
    def payoff(self, spot):
        return self.__payoff(self, spot)
        
def call_payoff(option, spot):
    return np.maximum(spot - option.strike, 0.0)

def put_payoff(option, spot):
    return np.maximum(option.strike - spot, 0.0)
    
def AmericanBinomial(option, spot, rate, volatility, dividend, steps):
    return np.maximum()
    
spot = 41
strike = 40.0
expiry = 1.0
rate = 0.08
volatility = 0.3
dividend = 0.0
steps = 3
num_nodes = steps + 1

dt = expiry / steps
u = np.exp((rate * dt) + volatility * np.sqrt(dt))
d = np.exp((rate * dt) - volatility * np.sqrt(dt))
pu = (np.exp(rate * dt) - d) / (u - d)
pd = 1 - pu
disc = np.exp(-rate * dt)
dpu = disc * pu
dpd = disc * pd
    
cashFlow_t = np.zeros((num_nodes, ))
spot_t = np.zeros((num_nodes, ))
option = VanillaOption(strike, expiry, call_payoff)
    
for i in range(num_nodes):
    spot_t[i] = spot * (u ** (steps - i)) * (d ** i)
    cashFlow_t[i] = option.payoff(spot_t[i]) 

for i in range((steps - 1), -1, -1):
    for j in range(i+1):
        cashFlow_t[j] = dpu * cashFlow_t[j] + dpd * cashFlow_t[j+1]
        spot_t[j] = spot_t[j] / u
        cashFlow_t[j] = np.maximum(cashFlow_t[j], option.payoff(spot_t[j]))    
    
price = cashFlow_t[0]

#The Call option price is:    
print ("The Option Price is: {0:.4f}".format(price))

The Option Price is: 7.0739


<span style="color:red">**Answer to Problem 2**</span>
 
European put option price is $\$2.99$, American put option price is $\$3.29$.

European call option price is $\$7.07$, American call option price is $\$7.07$.

The main difference between the american and the european solutions is that the American put option is worth more than the european put option. This is the case since it is beneficial to the holder of the option to excerise the option early, that is why the american is worth more. As we can see, the european and american call options are valued the same since early option exercise is not beneficial. I believe that the american call option would be more valuable than the european if there were dividends.</span>


The following code, I tried to vectorize and it didn't work well but I thought you would like to see it!

In [10]:
import numpy as np

def AmericanBinomialPricer(type, spot, strike, rate, vol, div, expiry, steps, american="false"):
   
   h = float(expiry) / steps
   u = np.exp((rate - div) * h + vol * np.sqrt(h))
   d = np.exp((rate - div) * h - vol * np.sqrt(h))
   nodes = steps + 1

   print(u,d)
   #We need to init the arrays using numpy so we can use vectors
   fs =  np.asarray([0.0 for i in range(nodes)])
       
   #Stock tree for calculations of expiration values
   fs2 = np.asarray([(spot * d**j * u**(steps - j)) for j in range(nodes)])
   print(fs2)
   #Vectorize the strikes
   fs3 =np.asarray( [float(strike) for i in range(nodes)])

   a = np.exp((rate-div) * h)
   pstar = (a - d)/ (u - d)
   oneMinusP = 1.0 - pstar

 
   if type =="C":
       fs = np.maximum(fs2-fs3, 0.0)
   else:
       fs = np.maximum(-fs2+fs3, 0.0)
   
   for i in range(steps, 0, -1):
      fstemp=np.exp(-rate * h ) * (pstar * fs[:-1] + oneMinusP * fs[1:])
      print(fstemp)
      fs2[:]=fs2[:]/d
      fs2 = np.delete(fs2, 0)
      fs3 = np.delete(fs3, 0)
      print(fstemp, fs2)
     
      if american=='true':
          #Check if the option is worth anything
          if type =="C":
               fs=np.maximum(fstemp,np.maximum(fs2-fs3,0))
          else:
               fs=np.maximum(fstemp,np.maximum(-fs2+fs3,0))
   return fs[0]

result = AmericanBinomialPricer("P", 41.0, 40.0, 0.08, 0.3, 0, 1, 3, american="true")
result

1.22124612015 0.863692553734
[ 74.6781323   52.81404438  37.3512727   26.41565494]
[ 0.          1.40091085  8.36287205]
[ 0.          1.40091085  8.36287205] [ 61.14912553  43.24602839  30.58455792]
[ 0.74094121  5.60292941]
[ 0.74094121  5.60292941] [ 50.07109093  35.4113947 ]
[ 3.29294759]
[ 3.29294759] [ 41.]


3.2929475853878247

**Answer to Problem 2:**


**Problem 3.**

In this proble, again utilizing the code written above, you will implement the Naive Monte Carlo Option Pricing Model for European puts and calls. Reprice both options above.

Recall, that if we can simulate prices under the risk-neutral density, then we can statistically estimate the option price as follows:

$$
\hat{C_{0}} = \frac{1}{M} \sum\limits_{j=1}^{M} C_{T,j}
$$

where $C_{T,j}$ is the option payoff at $t=T$ for the $j^{th}$ simulated spot price. This applies to both puts and calls, so again you can use polymorphism. 

Recall, that the first model that we will use to simulate prices in discrete time was called discrete-time Brownian motion, and can be implemented as follows:

$$
S_{t + h} = S_{t} \exp{ \left((r - \delta - \frac{1}{2} \sigma^{2}) + \sigma \sqrt{h} Z \right) }
$$

where $Z$ is a draw from a standard Normal distribution (i.e. $Z \sim N(0,1)$). And in the simple Black-Scholes world we set $t + h = T$. 

This gives us the estimated mean option price. We can also calculate the estimated standard error as follows:

$$
SE(\hat{C_{0}}) = \frac{SD(C_{0,j})}{\sqrt{M}}
$$

where $SD(C_{0,j})$ is the sample standard deviation, and is given as follows:

$$
SD(C_{0.j}) = \sqrt{\frac{1}{M-1} \sum\limits_{j=1}^{M} (C_{0,j} - \hat{C_{0}})^{2}}
$$

where $C_{0,j} = \exp{(-(r - \delta) T)} C_{T,j}$ is just the present value of the option payoff for simulated spot price $j$.

Report both $\hat{C_{0}}$ and $SE(\hat{C_{0}})$.

In [15]:
import numpy as np

class VanillaOption(object):
    def __init__(self, strike, expiry, payoff):
        self.__strike = strike
        self.__expiry = expiry
        self.__payoff = payoff
        
    @property
    def strike(self):
        return self.__strike
    
    @property
    def expiry(self):
        return self.__expiry
    
    def payoff(self, spot):
        return self.__payoff(self, spot)

def call_payoff(option, spot):
    return np.maximum(spot - option.strike, 0.0)

def put_payoff(option, spot):
    return np.maximum(option.strike - spot, 0.0)

def NaiveMonteCarlo(option, spot, rate, volatility, dividend, reps):
    expiry = option.expiry
    strike = option.strike
    dt = expiry
    nudt = (rate - dividend - 0.5 * volatility * volatility) * dt
    sigdt = volatility * np.sqrt(dt)
    z = np.random.normal(size=reps)

    spot_t = 0.0
    cashFlow_t = 0.0

    spot_t = spot * np.exp(nudt + sigdt * z)
    cashFlow_t = option.payoff(spot_t)

    price = cashFlow_t.mean() * np.exp(-rate * expiry)
    stdError = cashFlow_t.std() / np.sqrt(reps)

    return (price, stdError)

def main():
    strike = 40.0
    spot = 41.0
    rate = 0.08
    volatility = 0.30
    dividend = 0.0
    expiry = 1.0
    reps = 5000000

    the_option = VanillaOption(strike, expiry, call_payoff)
    option_price, stdError = NaiveMonteCarlo(the_option, spot, rate, volatility, dividend, reps)
    print("The call price is {0:.3f}, std error({1:.4f}) for {2} reps.".format(option_price, stdError, reps))

if __name__ == "__main__":
    main()

The call price is 6.959, std error(0.0048) for 5000000 reps.


In [13]:
import numpy as np

class VanillaOption(object):
    def __init__(self, strike, expiry, payoff):
        self.__strike = strike
        self.__expiry = expiry
        self.__payoff = payoff
        
    @property
    def strike(self):
        return self.__strike
    
    @property
    def expiry(self):
        return self.__expiry
    
    def payoff(self, spot):
        return self.__payoff(self, spot)

def call_payoff(option, spot):
    return np.maximum(spot - option.strike, 0.0)

def put_payoff(option, spot):
    return np.maximum(option.strike - spot, 0.0)

def NaiveMonteCarlo(option, spot, rate, volatility, dividend, reps):
    expiry = option.expiry
    strike = option.strike
    dt = expiry
    nudt = (rate - dividend - 0.5 * volatility * volatility) * dt
    sigdt = volatility * np.sqrt(dt)
    z = np.random.normal(size=reps)

    spot_t = 0.0
    cashFlow_t = 0.0

    spot_t = spot * np.exp(nudt + sigdt * z)
    cashFlow_t = option.payoff(spot_t)

    price = cashFlow_t.mean() * np.exp(-rate * expiry)
    stdError = cashFlow_t.std() / np.sqrt(reps)

    return (price, stdError)

def main():
    strike = 40.0
    spot = 41.0
    rate = 0.08
    volatility = 0.30
    dividend = 0.0
    expiry = 1.0
    reps = 5000000

    the_option = VanillaOption(strike, expiry, put_payoff)
    option_price, stdError = NaiveMonteCarlo(the_option, spot, rate, volatility, dividend, reps)
    print("The put option price is {0:.3f}, standard error({1:.4f}) for {2} reps.".format(option_price, stdError, reps))

if __name__ == "__main__":
    main()

The put option price is 2.886, standard error(0.0022) for 5000000 reps.


**Problem 4.**

In this problem you will write object-oriented code to simulate the ***Deck of Cards Workout*** from Ross Enamait's book *Never Gymless*. 

The details are as given as follows from Enamait's book:

This workout requires a standard deck of 52 playing cards. Each red card (diamonds and hearts) will require a set of burpees. Each black card (spades and clubs) will require a set of pushups. 

To perform the workout, start with a fully shuffled deck of cards. All face cards (Jacks, Queen, and King) have a value of 10. Aces have a value of 11. Number cards will be face value (ex. 7 of spades = 7 pushups). Do not use Joker cards for this problem.

Each card requires a set of pushups or burpees, depending on the color of the card. Strive to work through the entire deck as fast as possible. If burpees become too difficult, switch to squat thrusts. 

**Deck of Cards**

- Burpees for every red card
- Pushups for every black card

You will need to create classes with the appropriate data members and methods to simulate this process. You will need classes for the deck of cards, for the different suites of cards (diamonds, hearts, spades, clubs). You will need methods for drawing from the deck without replacement (*hint:* represent the deck as a list of cards, and use a pseudo-random number to sample randomly and the *pop* method to pull it from the list). You will also want a method that can randomly shuffle the deck, etc. Think carefully about this and plan it out before you begin to write any code. 

In [17]:
import numpy as np

def main():          
   
    a = np.random.choice(52,52,False)
    b = [11,2,3,4,5,6,7,8,9,10,10,10,10]
    c = ['A',2,3,4,5,6,7,8,9,10,'J','Q','K']
    pointer = 51
    Play = True
    while pointer > 0 and Play:
        if a[pointer] % 4 == 0:
            suit = 'Spades'
        elif a[pointer] % 4 == 1:
            suit = 'Hearts'
        elif a[pointer] % 4 == 2:
            suit = 'Clubs'
        elif a[pointer] % 4 == 3:
            suit = 'Diamonds'
           
        if a[pointer] % 2 == 0:
            exercise = 'Pushups'
        if a[pointer] % 2 == 1:
            exercise = 'Burpees'
           
        print('You got a', c[a[pointer]%13], 'of', suit,', Do',b[a[pointer]%13],exercise,'!')
        chara = input('Do you want to draw another card?(Y/N):')
        if chara == 'N':
            Play = False
        pointer = pointer -1

if __name__ == '__main__':
    main()

You got a 4 of Spades , Do 4 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 3 of Diamonds , Do 3 Burpees !
Do you want to draw another card?(Y/N):Y
You got a J of Spades , Do 10 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 5 of Spades , Do 5 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 6 of Clubs , Do 6 Pushups !
Do you want to draw another card?(Y/N):Y
You got a K of Clubs , Do 10 Pushups !
Do you want to draw another card?(Y/N):Y
You got a A of Spades , Do 11 Pushups !
Do you want to draw another card?(Y/N):YY
You got a 9 of Spades , Do 9 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 10 of Spades , Do 10 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 7 of Hearts , Do 7 Burpees !
Do you want to draw another card?(Y/N):Y
You got a 2 of Clubs , Do 2 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 5 of Clubs , Do 5 Pushups !
Do you want to draw another card?(Y/N):Y
You got a 5 of Diamonds , Do 5 Burpe